feat: add database dropdown to dashboard import

mistercrunch opened the pull request:
<img width="721" alt="Screen Shot 2020-06-20 at 8 45 56 AM" src="https://user-images.githubusercontent.com/487433/85205799-7a1fc480-b2d2-11ea-9257-d030a7a21cc9.png">

Currently, when importing a database from a JSON file, the process
looks at the database name from the source (the info is in the file)
and matches the datasources to that name. If no database by that name
exists, it simply fails.

With this PR, we add a database dropdown that allows the user to specify
which databases the datasources should target as the get upserted.

I want to stress that the code in this area is not in a great shape,
and that the challenge of serializing/deser the nested objects is
challenging, but that there should be a much better way to do this.
One of the improvement (out of scope for this PR) that would allow to
simplify those import/export would be to use UUIDs for
importable/exportable objects.

Another identified issue is the indirections between
`utils/import_expor_{model}.py` on top of `{Model}.import_object`. Not
addressing that here.

Next topic is the MVC stuff. Decided to stick with it for now as this is
more of a [obious missing feat:] than a rewrite.

### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->

### TEST PLAN
<!--- What steps should be taken to verify the changes -->

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

mistercrunch committed:
feat: add database dropdown to dashboard import

Currently, when importing a database from a JSON file, the process
looks at the database name from the source (the info is in the file)
and matches the datasources to that name. If no database by that name
exists, it simply fails.

With this PR, we add a database dropdown that allows the user to specify
which databases the datasources should target as the get upserted.

I want to stress that the code in this area is not in a great shape,
and that the challenge of serializing/deser the nested objects is
challenging, but that there should be a much better way to do this.
One of the improvement (out of scope for this PR) that would allow to
simplify those import/export would be to use UUIDs for
importable/exportable objects.

Another identified issue is the indirections between
`utils/import_expor_{model}.py` on top of `{Model}.import_object`. Not
addressing that here.

Next topic is the MVC stuff. Decided to stick with it for now as this is
more of a [obious missing feat:] than a rewrite.

mistercrunch commented on superset-frontend/src/components/FlashProvider.tsx:
```diff
@@ -42,7 +43,9 @@ class FlashProvider extends React.PureComponent<Props> {
       const [type, text] = message;
       const flash = flashObj[type];
-      this.props[flash](text);
+      if (this.props[flash]) {
```
Caught this bug as bycatch, this fails for the commonly used "alert" flashes.

mistercrunch commented on superset/models/dashboard.py:
```diff
@@ -311,6 +311,10 @@ def alter_positions(
         slices = copy(dashboard_to_import.slices)
+
+        # Clearing the slug to avoid conflicts
+        dashboard_to_import.slug = None
```
This was raising when loading the same dashboard (with a slug) twice. Currently the authority on whether to insert or update is the `remote_id` and didn't want to tangle the slug into that. Imported dashboards loose their slugs. We could lookup whether there will be a conflict or not and keep the slug / suffix it, but for now I'm just fixing this bug.

villebro replied:
Removing the slug might be unexpected behavior for the end user. I wonder if we should raise an alert toast if the slug has been removed until we figure out a more permanent solution?

mistercrunch replied:
So let's say I raise `Slug already exists!` user should do what? Go to change the slug from the source and re-export? I feel like we'd need to wizard that up and be like "Want to overwrite or create a new one?". Currently it'd be hard to do that.

I guess the form could offer "overwrite based on slug" checkbox or similar... ?

villebro replied:
TBH this is one of those things where it'd be really valuable to get usage statistics on how much slugs are used, how often slugged dashboards are exported + the slug is expected to work upon import. My guess is not so often, but who knows. But for the time being I'm comfortable with clearing the slug to avoid getting errors on repeated imports (idempotency FTW).

mistercrunch commented on superset/models/dashboard.py:
```diff
@@ -332,8 +336,8 @@ def alter_positions(
             # update json metadata that deals with slice ids
-            new_slc_id_str = "{}".format(new_slc_id)
-            old_slc_id_str = "{}".format(slc.id)
+            new_slc_id_str = str(new_slc_id)
```
Cruft, previous syntax was a safe py2/3 way to stringify... NOTE: there are a dozen other instances of this in the code base `git grep '"{}"\.format'`

mistercrunch commented on superset/utils/log.py:
```diff
@@ -168,6 +168,10 @@ def log(  # pylint: disable=too-many-locals
-        sesh = current_app.appbuilder.get_session
-        sesh.bulk_save_objects(logs)
-        sesh.commit()
+        try:
```
I hit this interesting bug where if the sqlalalchemy session is left in a bad state, this can raise. I feel like the right/easy thing to do here is to silence this. Failing at logging an analytic event should never fail the app as hard as this did.

mistercrunch reviewed (commented):
Some PR notes

mistercrunch committed:
isort \!? 0%^$%Y$&?%$^?%0^?

mistercrunch committed:
fix tests

eugeniamz commented:
Do you think that doing this will resolve the issue that if Dashboard to import is using legacy Druid Driver?

mistercrunch commented:
@eugeniamz we're more heading towards deprecation at this time, I'd rather help people migrate off the legacy connector and onto SQL druid than investing in that direction.

In theory we could make the dropdown include databases + druid clusters, but that's one more thing to untangle/deprecate...

villebro commented on superset/connectors/sqla/models.py:
```diff
@@ -1218,7 +1218,10 @@ def fetch_metadata(self, commit: bool = True) -> None:
     def import_obj(
-        cls, i_datasource: "SqlaTable", import_time: Optional[int] = None
+        cls,
+        i_datasource: "SqlaTable",
```
Not yours, but the name `i_datasource` makes me uncomfortable 😬

mistercrunch replied:
"input" ?

mistercrunch replied:
Let's leave that to another PR

villebro reviewed (commented):
A few comments.

dpgaspar commented on superset/utils/log.py:
```diff
@@ -168,6 +168,10 @@ def log(  # pylint: disable=too-many-locals
+            sesh = current_app.appbuilder.get_session
+            sesh.bulk_save_objects(logs)
+            sesh.commit()
+        except Exception as ex:  # pylint: disable=broad-except
```
`except SQLAlchemyError as ex:`

dpgaspar commented on superset/views/core.py:
```diff
@@ -538,8 +539,13 @@ def import_dashboards(self) -> FlaskResponse:
         """Overrides the dashboards using json instances from the file."""
         f = request.files.get("file")
         if request.method == "POST" and f:
+            success = False
```
nit: replace `f` with `import_file` or `dashboard_file`

mistercrunch committed:
pre-committing to py3.6

mistercrunch committed:
address dpgaspar's comments

mistercrunch commented:
@dpgaspar @villebro , addressed most comments. Clearly export/import needs more attention, but this tackles important shortcoming of lacking the ability to target a database, and fixes the slug bug.

willbarrett commented:
FYI @mistercrunch this needs a rebase.

mistercrunch committed:
revert isort
